zookeeper/src/main/java/com/linecorp/armeria/client/zookeeper/ZooKeeperEndpointGroup.java

```diff
@@ -59,66 +62,95 @@ public ZooKeeperEndpointGroup(String zkConnectionStr, String zNodePath, int sess
     /**
      * Create a ZooKeeper-based {@link EndpointGroup}, endpoints will be retrieved from a node's all children's
      * node value using {@link NodeValueCodec}.
+     *
      * @param zkConnectionStr a connection string containing a comma separated list of {@code host:port} pairs,
      *                        each corresponding to a ZooKeeper server
      * @param zNodePath       a zNode path e.g. {@code "/groups/productionGroups"}
-     * @param sessionTimeout  Zookeeper session timeout in milliseconds
-     * @param nodeValueCodec  the nodeValueCodec
+     * @param sessionTimeout  ZooKeeper session timeout in milliseconds
+     * @param nodeValueCodec  the {@link NodeValueCodec}
      */
     public ZooKeeperEndpointGroup(String zkConnectionStr, String zNodePath, int sessionTimeout,
                                   NodeValueCodec nodeValueCodec) {
-        zooKeeperConnector = new ZooKeeperConnector(zkConnectionStr, zNodePath, sessionTimeout,
-                                                    createListener());
+        requireNonNull(zNodePath, "zNodePath");
+        checkArgument(!zNodePath.isEmpty(), "zNodePath can't be empty");
+        checkArgument(sessionTimeout > 0, "sessionTimeoutMillis: %s (expected: > 0)",
+                      sessionTimeout);
         this.nodeValueCodec = requireNonNull(nodeValueCodec, "nodeValueCodec");
-        zooKeeperConnector.connect();
-    }
-
-    @Override
-    public void close() {
-        zooKeeperConnector.close(true);
+        this.internalClient = true;
+        this.client = CuratorFrameworkFactory.builder()
+                                             .connectString(zkConnectionStr)
+                                             .retryPolicy(ZooKeeperDefaults.DEFAULT_RETRY_POLICY)
+                                             .sessionTimeoutMs(sessionTimeout)
+                                             .build();
+        client.start();
+        boolean success = false;
+        try {
+            pathChildrenCache = pathChildrenCache(zNodePath);
+            pathChildrenCache.start();
+            success = true;
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
+        } finally {
+            if (!success) {
+                client.close();
+            }
+        }
     }
 
     /**
-     * Create a {@link ZooKeeperListener} listens specific ZooKeeper events.
-     * @return  {@link ZooKeeperListener}
+     * Create a ZooKeeper-based {@link EndpointGroup}, endpoints will be retrieved from a node's all children's
```

Comment:
Creates?

Reply:
All the other constructors have been used 'Create'. Moreover, other similar cases also use 'Create', not 'Creates'.

It seems like we need some kind of guideline here.

Reply:
All the other constructors has been used 'Create'. Moreover, other similar cases also use 'Create', not 'Creates'.

It seems like we need some kind of guideline here...

Reply:
Subject verb agreement is one of the thing that I was confused when writing documents. :-)
We start the description with a 3rd person singular verb. I'm not sure why it is. But I think if you put the name of this constructor of method, at the beginning of the sentence, that makes sense. 
But we are not so strict to them like other descriptions in this class. :-)


But like you see other constructors in this class, it's no big deal.

Reply:
It's like:

    (This method) creates ...

which becomes:

    Creates ...